adapters/repos/db/lsmkv/segment_net_count_additions.go

```diff
@@ -0,0 +1,180 @@
+
+func (ind *segment) countNetPath() string {
+	return countNetPathFromSegmentPath(ind.path)
+}
+
+func countNetPathFromSegmentPath(segPath string) string {
+	extless := strings.TrimSuffix(segPath, filepath.Ext(segPath))
+	return fmt.Sprintf("%s.cna", extless)
+}
+
+func (ind *segment) initCountNetAdditions(exists existsOnLowerSegmentsFn) error {
+	if ind.strategy != SegmentStrategyReplace {
+		// replace is the only strategy that supports counting
+		return nil
+	}
+
+	ok, err := fileExists(ind.countNetPath())
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		err = ind.loadCountNetFromDisk()
+		if err == nil {
+			return nil
+		}
+
+		if err != ErrInvalidChecksum {
+			// not a recoverable error
+			return err
+		}
+
+		// now continue re-calculating
+	}
+
+	var lastErr error
+	countNet := 0
+	cb := func(key []byte, tombstone bool) {
+		existedOnPrior, err := exists(key)
+		if err != nil {
+			lastErr = err
+		}
+
+		if tombstone && existedOnPrior {
+			countNet--
+		}
+
+		if !tombstone && !existedOnPrior {
+			countNet++
+		}
+	}
+
+	extr := newBufferedKeyAndTombstoneExtractor(ind.contents, ind.dataStartPos,
+		ind.dataEndPos, 10e6, ind.secondaryIndexCount, cb)
+
+	extr.do()
+
+	ind.countNetAdditions = countNet
+
+	if lastErr != nil {
+		return lastErr
+	}
+
+	if err := ind.storeCountNetOnDisk(); err != nil {
+		return fmt.Errorf("store count net additions on disk: %w", err)
+	}
+
+	return nil
+}
+
+func (ind *segment) storeCountNetOnDisk() error {
+	return storeCountNetOnDisk(ind.countNetPath(), ind.countNetAdditions)
+}
+
+// prefillCountNetAdditions is a helper function that can be used in
+// compactions. A compacted segment behaves exactly as the two segments it
+// replaces. As a result the count net additions of a compacted segment is
+// simply the sum of the old two segments.
+//
+// by "prefilling" which means creating the file on disk, the subsequent
+// newSegment() call can skip re-calculating the count net additions which
+// would have a high cost on large segment groups.
+func prefillCountNetAdditions(segPath string, updatedCountNetAdditions int) error {
+	return storeCountNetOnDisk(countNetPathFromSegmentPath(segPath),
+		updatedCountNetAdditions)
+}
+
+func storeCountNetOnDisk(path string, value int) error {
+	buf := new(bytes.Buffer)
+
+	if err := binary.Write(buf, binary.LittleEndian, uint64(value)); err != nil {
+		return fmt.Errorf("write cna to buf: %w", err)
+	}
+
+	data := buf.Bytes()
+	chksm := crc32.ChecksumIEEE(data)
+
+	f, err := os.Create(path)
+	if err != nil {
+		return fmt.Errorf("open file for writing: %w", err)
+	}
+
+	if err := binary.Write(f, binary.LittleEndian, chksm); err != nil {
+		return fmt.Errorf("write checkusm to file: %w", err)
+	}
+
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("write cna data to file: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("close cna file: %w", err)
+	}
+
+	return nil
+}
```

Comment:
This looks identical to the function for bloom filters. If it is not too much work, can you combine this?

Reply:
Done